server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -551,8 +553,8 @@ protected SendSnapshotResult sendSnapshot(final long startingSeqNo, long require
             logger.trace("no translog operations to send");
         }
 
-        final CancellableThreads.IOInterruptable sendBatch =
-                () -> targetLocalCheckpoint.set(recoveryTarget.indexTranslogOperations(operations, expectedTotalOps));
+        final CancellableThreads.IOInterruptable sendBatch = () ->
+            targetLocalCheckpoint.set(recoveryTarget.indexTranslogOperations(operations, expectedTotalOps, shard.getMaxAutoIdTimestamp()));
```

Comment:
Instead of using a new one for every batch that is to be sent, I would prefer to capture this after we call `cancellableThreads.execute(() -> shard.waitForOpsToComplete(endingSeqNo));` in RecoverySourceHandler, and then only pass that same value. You could also add a comment then and there saying why we do it.